sympy/printing/smtlib.py

```diff
@@ -0,0 +1,515 @@
+    known_functions : dict, optional
+        A dictionary where keys are ``Function``, ``Relational``, ``BooleanFunction``, or ``Expr`` instances and values are their SMT string representations.
+        If not given, a partial listing optimized for dReal solver (but compatible with others) will be used.
+    known_constants: dict, optional
+        A dictionary where keys are ``NumberSymbol`` instances and values are their SMT variable names.
+        When using this feature, extra caution must be taken to avoid naming collisions between user symbols and listed constants.
+        If not given, constants will be expanded inline i.e. '3.14159' instead of 'MY_SMT_VARIABLE_FOR_PI'.
+    prefix_expressions: list, optional
+        A list of lists of ``str`` and/or expressions to convert into SMTLib and prefix to the output.
+    suffix_expressions: list, optional
+        A list of lists of ``str`` and/or expressions to convert into SMTLib and postfix to the output.
+    log_warn: lambda function, optional
+        A function to record all warnings during potentially risky operations.
+        Soundness is a core value in SMT solving, so it is good to log all assumptions made.
+        If not given, builtins ``print`` will be used.
+
+    Examples
+    ========
+
+    >>> noop = (lambda _: None)
+    >>> from sympy import smtlib_code, symbols, sin, Eq
+    >>> x = symbols('x')
+    >>> smtlib_code(sin(x).series(x).removeO(), log_warn=noop)
+    '(declare-const x Real)\n(+ x (* (/ -1 6) (pow x 3)) (* (/ 1 120) (pow x 5)))'
+
+    >>> from sympy import Rational
+    >>> x, y, tau = symbols("x, y, tau")
+    >>> smtlib_code((2*tau)**Rational(7, 2), log_warn=noop)
+    '(declare-const tau Real)\n(* 8 (pow 2 (/ 1 2)) (pow tau (/ 7 2)))'
+
+    ``Piecewise`` expressions are implemented with ``ite`` expressions by default.
+    Note that if the ``Piecewise`` lacks a default term, represented by
+    ``(expr, True)`` then an error will be thrown.  This is to prevent
+    generating an expression that may not evaluate to anything.
+
+    >>> from sympy import Piecewise
+    >>> pw = Piecewise((x + 1, x > 0), (x, True))
+    >>> smtlib_code(Eq(pw, 3))
+    Could not infer type of `x`. Defaulting to float.
+    '(declare-const x Real)\n(assert (= (ite (> x 0) (+ 1 x) x) 3))'
+
+    Custom printing can be defined for certain types by passing a dictionary of
+    PythonType : "SMT Name" to the ``known_types``, ``known_constants``, and ``known_functions`` kwargs.
+
+    >>> from typing import Callable
+    >>> from sympy import Function, Add
+    >>> f = Function('f')
+    >>> g = Function('g')
+    >>> smt_builtin_funcs = {  # functions our SMT solver will understand
+    ...   f: "existing_smtlib_fcn",
+    ...   Add: "sum",
+    ... }
+    >>> user_def_funcs = {  # functions defined by the user must have their types specified explicitly
+    ...   g: Callable[[int], float],
```

Comment:
I'm not familiar enough with smtlib, but are the Python typing types expressive enough to represent everything we might want?

Reply:
The answer is likely no, but that shouldn't matter because the types which might not be representable aren't really relevant to sympy nor this 1-way printer binding.

Reply:
@soonhokong @scungao I'm not super familiar with SMT type systems— do you think this current implementation (directly mapping Python types to SMT types) should suffice? For now, I only support:

| Python type | SMT type |
| ------------ | ---------- |
| `bool` | `Bool` |
| `int` | `Int` |
| `float` | `Real` |
| `Callable[...any permutation of the above...]` | UFs over the above |